### PR TITLE
Use keyboard scanner

### PR DESCRIPTION
**Pull request by andreievg**
Please see README.md for full description of work, and the reasoning.

Apart from adding keyboard barcode scanner in electron app, also added a toggle in admin page to use either keyboard or serial scanner (it's technically not serial usb yet, but done this way in anticipation). And added ability to switch either keyboard or serial mode on electron end

### Testing

In electron client app:

Switch Zebra scanner to keyboard mode:

Screenshot 2023-05-20 at 12 51 12 AM

In Admin page change to keyboard scanner mode.

In stock view try scanning some barcode (see batch and expiry reflect in stock correctly), there is a complex barcode in README.md and this one: 

Save one of the barcode against stock line

Start scanning in outbound shipment, make sure navigation to correct item and batch is done and make sure keyboard works correctly (when entering quantities, tabbing pressing enter and when searching for item when barcode is not linked).

Now change back to USB serial mode and repeat the task (making sure USB serial still works correctly), here is a barcode to use for USB serial mode  **NOTE** it is not usb serial mode, but just named this way in anticipation of changing to using USB serial.

Screenshot 2023-05-20 at 4 47 59 PM

### CODE REVIEW

File: client/packages/common/src/hooks/useNativeClient/types.ts [hunk 1]
```diff
@@ -66,3 +68,5 @@ export type BarcodeScanner = {
   product?: string;
   connected: boolean;
 };
+
+export type ScannerType = 'usb_serial' | 'usb_keyboard';
```
Comment (andreievg) [inline client/packages/common/src/hooks/useNativeClient/types.ts:72]: used string constants, since it's slightly easier to use when only a few variations (it's still just as much type safety)

File: client/packages/common/src/utils/BarcodeScannerContext.tsx [hunk 2]
```diff
@@ -84,6 +86,7 @@ export const BarcodeScannerProvider: FC<PropsWithChildrenOnly> = ({
   const { error } = useNotification();
   const { electronNativeAPI } = window;
   const [scanner, setScanner] = useState<BarcodeScanner | null>(null);
+  const [scannerType, _setScannerType] = useState<ScannerType>('usb_serial');
```
Comment (andreievg) [inline client/packages/common/src/utils/BarcodeScannerContext.tsx:89]: wasn't sure how to name `inner` setScannerType, so prefixed with `_` ¯\_(ツ)_/¯

File: client/packages/common/src/utils/BarcodeScannerContext.tsx [hunk 3]
```diff
@@ -193,9 +196,15 @@ export const BarcodeScannerProvider: FC<PropsWithChildrenOnly> = ({
     }
   };
 
+  const setScannerType = (type: ScannerType) => {
+    electronNativeAPI.setScannerType(type);
+    electronNativeAPI?.linkedBarcodeScannerDevice().then(setScanner);
```
Comment (andreievg) [inline client/packages/common/src/utils/BarcodeScannerContext.tsx:201]: have to call linkedBarcodeScannerDevice, to get and set scanner (as connected or not connected, keyboardMode return connected scanner with some defaults for non optional fields of BarcodeScanner)

File: client/packages/electron/src/electron.ts [hunk 4]
```diff
@@ -7,15 +7,18 @@ import {
...
 import HID from 'node-hid';
 import ElectronStore from 'electron-store';
+import { KeyboardScanner } from './keyboardScanner/keyboardScanner';
 
 const SERVICE_TYPE = 'omsupply';
 const PROTOCOL_KEY = 'protocol';
 const CLIENT_VERSION_KEY = 'client_version';
 const HARDWARE_ID_KEY = 'hardware_id';
 const BARCODE_SCANNER_DEVICE_KEY = 'barcode_scanner_device';
+const SCANNER_TYPE = 'scanner_type';
```
Comment (andreievg) [inline client/packages/electron/src/electron.ts:21]: scanner type preference is stored in electron app data

File: client/packages/electron/src/electron.ts [hunk 5]
```diff
@@ -28,9 +31,11 @@ type StoreType = {
 class Scanner {
   device: HID.HID | undefined;
   barcodeScanner: BarcodeScanner | undefined;
+  window: BrowserWindow;
 
-  constructor() {
+  constructor(window: BrowserWindow) {
```
Comment (andreievg) [inline client/packages/electron/src/electron.ts:36]: added to constructor rather then start(), to allign with KeyboardScanner

File: client/packages/electron/src/electron.ts [hunk 6]
```diff
@@ -117,7 +122,6 @@ class Scanner {
 
 const store = new ElectronStore<StoreType>();
 const discovery = new dnssd.Browser(dnssd.tcp(SERVICE_TYPE));
-const barcodeScanner = new Scanner();
```
Comment (andreievg) [inline client/packages/electron/src/electron.ts:120]: moved to app start()

File: client/packages/electron/src/electron.ts [hunk 7]
```diff
@@ -184,22 +188,39 @@ const start = (): void => {
   );
 
   ipcMain.handle(IPC_MESSAGES.CONNECTED_SERVER, async () => connectedServer);
-  ipcMain.handle(IPC_MESSAGES.START_BARCODE_SCAN, () =>
```
Comment (andreievg) [inline client/packages/electron/src/electron.ts:187]: moved below to one common barcode section, and instead of barcodeScaner, using getCurrentScanner()

File: client/packages/electron/src/electron.ts [hunk 8]
```diff
@@ -184,22 +188,39 @@ const start = (): void => {
...
     const servers = discoveredServers;
     discoveredServers = [];
     return { servers };
   });
 
+  // BARCODES
+  const SerialScanner = new Scanner(window);
+  const keyboardScanner = new KeyboardScanner(window);
+  const getCurrentScanner = () =>
+    store.get(SCANNER_TYPE, 'usb_serial') == 'usb_serial'
```
Comment (andreievg) [inline client/packages/electron/src/electron.ts:202]: default to preferred/recommended scanner mode

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts
Comment (andreievg) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:1]: Logic of this file is described in README.md and in the comments

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+  buffer: KeyEvent[];
+  settings: KeyboardScannerSettings;
+
+  constructor(window: BrowserWindow) {
+    // this.device = this.findDevice();
+    this.window = window;
+    this.scanning = false;
+    this.lockBuffer = false;
+    this.allowEvents = false;
+    this.settings = {
```
Comment (andreievg) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:72]: I thought this was going to be a configurable setting, but decided not to add these configurations yet, but kept the data structure

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+    try {
+      this.lockBuffer = true;
+
+      let checkSequence: KeyEvent[] = [];
+      let passThrough: KeyEvent[] = [];
+      let barcode: KeyEvent[] = [];
+
+      this.buffer.forEach((current, index) => {
+        // Get ms for next input event or now
+        const nextMs = this.buffer[index + 1]?.ms ?? Date.now();
```
Comment (andreievg) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:109]: `now` because sequence is considered finished if it's maxMSBetweenScannerKeyInputs from `now`

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 11]
```diff
@@ -6,6 +6,7 @@ import {
   Box,
   LoadingButton,
   LocaleKey,
+  Switch,
```
Comment (andreievg) [inline client/packages/host/src/Admin/ElectronSettings.tsx:9]: Using switch, same as request edit modal (when selection order in quantity or packs)

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 12]
```diff
@@ -49,7 +50,11 @@ export const ElectronSettings = () => {
   const [scanner, setScanner] = useState<BarcodeScanner | null>(null);
   const [isScanning, setIsScanning] = useState(false);
   const { error, success } = useNotification();
-  const { setScanner: setBarcodeScanner } = useBarcodeScannerContext();
+  const {
```
Comment (andreievg) [inline client/packages/host/src/Admin/ElectronSettings.tsx:53]: technically scanner and setScanner should come from context too

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 13]
```diff
@@ -84,7 +89,7 @@ export const ElectronSettings = () => {
 
   useEffect(() => {
     electronNativeAPI?.linkedBarcodeScannerDevice().then(setScanner);
-  }, [electronNativeAPI]);
+  }, [electronNativeAPI, scannerType]);
```
Comment (andreievg) [inline client/packages/host/src/Admin/ElectronSettings.tsx:92]: when scannerType is refreshed we should update scanner (as connected etc..), this is already done in context, but didn't want to change too much

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 14]
```diff
@@ -93,47 +98,73 @@ export const ElectronSettings = () => {
...
-              </Typography>
-            </Box>
-          ) : (
-            <Scanner scanner={scanner} />
-          )
-        }
-      />
-      <Setting
-        title=""
+        title={'Scanner Type'}
```
Comment (andreievg) [inline client/packages/host/src/Admin/ElectronSettings.tsx:103]: no changes to existing settings, just moved them behind `scannerType === 'usb_serial'`

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 15]
```diff
@@ -93,47 +98,73 @@ export const ElectronSettings = () => {
...
-              </Box>
-            )}
-            <Box display="flex" justifyContent="flex-end">
-              <LoadingButton onClick={startDeviceScan} isLoading={isScanning}>
-                {t('label.detect-scanners')}
-              </LoadingButton>
+          <Box display="flex" justifyContent="flex-end" alignItems="center">
+            <Switch
+              label={'USB Serial'}
+              checked={scannerType === 'usb_keyboard'}
```
Comment (andreievg) [inline client/packages/host/src/Admin/ElectronSettings.tsx:108]: This is weird (to have opposite in checked and label), but I think this is how the Switch between two values works in other places

File: client/packages/electron/src/electron.ts [hunk 16]
```diff
@@ -184,22 +188,39 @@ const start = (): void => {
...
+  );
   ipcMain.handle(IPC_MESSAGES.LINKED_BARCODE_SCANNER_DEVICE, async () =>
-    barcodeScanner.linkedScanner()
+    getCurrentScanner().linkedScanner()
+  );
+  ipcMain.handle(IPC_MESSAGES.START_BARCODE_SCAN, () =>
+    getCurrentScanner().start()
+  );
+  ipcMain.handle(IPC_MESSAGES.STOP_BARCODE_SCAN, () =>
+    getCurrentScanner().stop()
```
Comment (mark-prins) [inline client/packages/electron/src/electron.ts:217]: These, and GET_SCANNER_TYPE & SET_SCANNER_TYPE, could all be handled with one IPC message if
- the LINKED_BARCODE_SCANNER_DEVICE was replaced with a call to get BARCODE_SCANNER which called `getCurrentScanner` 
- and the methods on that class were called by the caller (BarcodeScannerContext in this case)
- the class handled the scanner type: added to the constructor and methods added to get/set

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 17]
```diff
@@ -0,0 +1,231 @@
...
+  'Delete',
+  'Clear',
+];
+
+// Scan Log
+const PROCESSOR_INTERVAL = 50;
+
+type KeyEvent = { input: Electron.Input; ms: number };
+
+// Some keys need to be removed (mention about the thig)
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:36]: Unsure what the last part means sorry: `(mention about the thig)`

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+  // this lock is to make sure it's edited syncrhonously
+  lockBuffer: boolean;
+  // In scanning mode, when non barcode data entry is made, should allow it (no preventDefault())
+  allowEvents: boolean;
+  scanning: boolean;
+  buffer: KeyEvent[];
+  settings: KeyboardScannerSettings;
+
+  constructor(window: BrowserWindow) {
+    // this.device = this.findDevice();
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:67]: not required?

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 19]
```diff
@@ -0,0 +1,231 @@
...
+        if (diff < this.settings.maxMSBetweenScannerKeyInputs) return;
+        // If exceeding maxMSBetweenScannerKeyInputs and sequence length
+        // is not long enough to be barcode, pass all input through
+        if (checkSequence.length < this.settings.minBarcodeLength) {
+          passThrough = [...passThrough, ...checkSequence];
+          checkSequence = [];
+        } else {
+          // If exceeding maxMSBetweenScannerKeyInputs but sequnce length
+          // is long enought to be barcode, record as barcode;
+          // Very unlikely that two barcodes will be scanned within PROCESSOR_INTERVAL
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:124]:
```suggestion
          // If exceeding maxMSBetweenScannerKeyInputs but sequence length
          // is long enough to be a barcode, record as barcode;
          // Very unlikely that two barcodes will be scanned within PROCESSOR_INTERVAL
```

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+          barcode = checkSequence;
+          checkSequence = [];
+        }
+      });
+
+      // Remember left over for next check
+      this.buffer = checkSequence;
+      if (barcode.length != 0) {
+        console.log(JSON.stringify(barcode, null, ' '));
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:133]: not required?

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+        passThrough.forEach(({ input: { key } }) => {
+          this.sendKey(key);
+        });
+
+        // Delay slightly to allow above events to go through, before unlocking buffer
+        await new Promise(resolve => setTimeout(resolve, 10));
+        this.allowEvents = false;
+      }
+    } catch (e) {
+      console.log(e);
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:150]: ?
```suggestion
      console.error(e);
```

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts [hunk 22]
```diff
@@ -0,0 +1,231 @@
...
+  bindInputEvent() {
+    // Bind keyboard capture on startup (this.scanning is false by default)
+    this.window.webContents.on('before-input-event', (event, input) => {
+      if (!this.scanning) return;
+      if (keysToPassThrough.includes(input.key)) return;
+
+      while (this.lockBuffer) {
+        // Buffer will be locked when processing input, and when input is being processed and key event
+        // is identified as normal keyboard event it will be triggered from processor, this event should be
+        // passed through (no preventDefault)
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/keyboardScanner.ts:166]: I'm finding the comment hard to follow. Some formatting might help - also unsure of what 'it' is and the 'event'
```   // Buffer will be locked when processing input
        // When input is being processed and key event is identified as normal keyboard event
        // it [what does 'it' mean here?] will be triggered from processor
        // this event [which event are we talking about?] should be passed through (no preventDefault) [hm?]
```

**Review by mark-prins — COMMENTED** [review]
untested - have just had a read through. Wasn't sure if this was to be reviewed yet or not, but thought some comments might be helpful?

File: client/packages/common/src/utils/BarcodeScannerContext.tsx (hunk 2 above)
Comment (andreievg) [reply client/packages/common/src/utils/BarcodeScannerContext.tsx]: Great, i like `local` in name, thanks: ec8f8bbddfb5c6c908e5529e88bd61cd30d5e17e

File: client/packages/host/src/Admin/ElectronSettings.tsx (hunk 12 above)
Comment (andreievg) [reply client/packages/host/src/Admin/ElectronSettings.tsx:53]: > the scanner type would be better coming from the context as well,

It is ; )

File: client/packages/host/src/Admin/ElectronSettings.tsx (hunk 15 above)
Comment (andreievg) [reply client/packages/host/src/Admin/ElectronSettings.tsx:108]: I made an issue: https://github.com/openmsupply/open-msupply/issues/1813

File: client/packages/electron/src/keyboardScanner/README.md [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+
+If sequence input is not `finalised` (still ongoing while processor is triggered), it will be put back into buffer.
+
+Processor is started (setInterval) when scanning operation is started and will be cleared when scanning operation is stopped (clearInterval)
+
+#### Passing through input
+
+Any controll keys (keysToPassThrough array) are passed through automatically (when input is captured). Other input that is identified by processor as not scanner input is passed through by manually calling keyDown and char event (apparently that is what [chromium](https://stackoverflow.com/a/53223619) needs to consider it input entered by keyboard). 
+
+Since we are capturing input with `window.webContents.on('before-input-event'` we have to temporary allow this input to go through with `allowEvents` flag.
```
[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/README.md]: 1f65bc485dff4825bce7160aa31f8858253a0c65, hopefully this makes more sense

File: client/packages/electron/src/keyboardScanner/README.md [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+#### Other notes
+
+Only `keyDown` input is captured in buffer to be processed, `keyUp` events are captured but are not appended to buffer.
+
+We also capture and ignore `isAutoRepeat` type input, this is when they key is held down (only the first keyDown even is captured and stored in buffer for processing).
+
+When barcode is created from input sequence, we only keep inputs starting with Digit and Key, and instead of using `key` of input, we use `code` of input, i.e. (Digit1 or KeyA), this means any special characters in barcode are ignored and capitalisation is ignored. This is done to make sure that keyboard OS settings dont' interfere with resulting barcode (i.e. when French AZERTY keyboard mode is on). This should only affect lot/batch number, and from my understanding those are always capital english letters.
+
+In more complex barcodes, as per image below, special character seperates lot and [serial number](https://www.databar-barcode.info/application-identifiers/), for Zebra this special character seems to be `Numpad0` + 'alt' + 'iskeypad' and for ZKTeco it was 'F8', this is replaced with charcode `29` (when scanned with USB serial mode this appears as charcode 29):
```
[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/README.md:43]: yeah, this whole fix is hacky, but the impact of this one in particular is quite low, without this change I was getting serial appended to batch, this is the only place it appears, for simple barcodes this is not an issue, only affect barcodes with batch, expiry AND serial (still not super happy about it)

File: client/packages/host/src/Admin/ElectronSettings.tsx [hunk 25]
```diff
@@ -93,47 +98,73 @@ export const ElectronSettings = () => {
...
             </Box>
           </Box>
         }
       />
+      {scannerType === 'usb_serial' && (
+        <>
+          <Setting
+            title={t('label.current')}
+            component={
+              !scanner ? (
```
[the remark this answers is not in the document]
Comment (andreievg) [reply client/packages/host/src/Admin/ElectronSettings.tsx]: bb691aa24d85f4543684c4394371ea29c7160d64, thanks, that does look much better

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 17 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts]: Oops, lol the thig was removed: fe936ba938a244f34dcdc05fb744b020bbf8d305

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 18 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts]: 30413c7e9aba96321acaaf08154fdce811d3832c

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 21 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts:149]: 4de4cfee028093782782141a22456d9381690c29, thanks

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 20 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts:132]: Hmm i kind of really want this one in, in case there is an issue can ask to open app with terminal and see what the barcode scan was, hmm, can I force console log in window (and maybe debug console log) ? Will try this

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 22 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts]: Hmm, yeah that looks very hard to read, I changed based on questions in the example comment: 4a1cac38cdfbd0f43d28b2c916ba29d9159866a6, is this ok ?

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 20 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts:132]: https://github.com/openmsupply/open-msupply/compare/Use-keyboard-scanner...Log-electron-to-console, maybe could do something like this ¯\_(ツ)_/¯ I am also ok with deleting this log, although I did find it useful, that's how I figured out about the special separator character

File: client/packages/electron/src/keyboardScanner/README.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/electron/src/keyboardScanner/README.md:43]: agree - the risk is low, and it's well documented.

File: client/packages/electron/src/keyboardScanner/README.md [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+
+If sequence input is not `finalised` (still ongoing while processor is triggered), it will be put back into buffer.
+
+Processor is started (setInterval) when scanning operation is started and will be cleared when scanning operation is stopped (clearInterval)
+
+#### Passing through input
+
+Any control keys (keysToPassThrough array) are passed through automatically (when input is captured). Other input that is identified by processor as not scanner input is passed through by manually calling keyDown and char event (apparently that is what [chromium](https://stackoverflow.com/a/53223619) needs to consider it input entered by keyboard). 
+
+Since we are capturing input with `window.webContents.on('before-input-event'` we have to temporary allow this input to go through with `allowEvents` flag. Which is set in the processor when input is identified as not being from scanner.
```
Comment (mark-prins) [inline client/packages/electron/src/keyboardScanner/README.md:33]: :nit 
```suggestion
Since we are capturing input with `window.webContents.on('before-input-event'` we have to temporary allow this input to go through with `allowEvents` flag which is set in the processor when the input is identified as not being from scanner.
```

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 20 above)
Comment (mark-prins) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts:132]: pushing it out - to log to the render console would be handy. it's not easy to see the electron console in production, but the web app / renderer console is ok. and sure - fine to leave in console logging that's useful / deliberate. Maybe add a comment, so that it isn't nit-picked in future? `// this console log is super useful and is supposed to be here!`

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 22 above)
Comment (mark-prins) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts]: sweet - I understand now! 😄

**Review by mark-prins — APPROVED** [review]
🥳  works like a charm!

In testing I've noticed that the stock lines in outbound shipments aren't being selected / focussed in the way I expect. That's not caused by this PR though, will need to be investigated elsewhere

File: client/packages/electron/src/keyboardScanner/keyboardScanner.ts (hunk 20 above)
Comment (andreievg) [reply client/packages/electron/src/keyboardScanner/keyboardScanner.ts:134]: Great idea, thanks: 4de4cfee028093782782141a22456d9381690c29

**Comment by andreievg** [comment]
@mark-prins thank you very much for your review. Even though this one doesn't have an associated issue, i'll merge it, it's been introduced to customer already (sorry, sneakily got it in, but I think it's been good)
